src/rdrand.c

```diff
@@ -31,33 +31,85 @@
 #include "rdtime.h"
 #include "tinycthread.h"
 #include "rdmurmur2.h"
+#ifndef _WIN32
+/* getentropy() can be present in one of these two */
+#include <unistd.h>
+#include <sys/random.h>
+#endif
 
-int rd_jitter(int low, int high) {
+/* Initial seed with time+thread id */
+unsigned int rd_seed() {
+        unsigned int seed = 0;
+        struct timeval tv;
+        rd_gettimeofday(&tv, NULL);
+        seed = (unsigned int)(tv.tv_usec);
+        seed ^= thrd_current_id();
+
+        /* Apply the murmur2 hash to distribute entropy to
+         * the whole seed. */
+        seed = (unsigned int)rd_murmur2(&seed, sizeof(seed));
+        return seed;
+}
+
+static int rd_rand() {
         int rand_num;
 #if HAVE_RAND_R
         static RD_TLS unsigned int seed = 0;
-
-        /* Initial seed with time+thread id */
         if (unlikely(seed == 0)) {
-                struct timeval tv;
-                rd_gettimeofday(&tv, NULL);
-                seed = (unsigned int)(tv.tv_usec);
-                seed ^= (unsigned int)(intptr_t)thrd_current();
-
-                /* When many threads are created at the same time and the
-                 * thread id is different only by a few bits it's possible that
-                 * `rand_r`, that is initially multiplying by `1103515245`,
-                 * truncates the variable bits and uses the same seed for
-                 * different threads. By applying `murmur2` we ensure that seed
-                 * variability is distributed across various bits at different
-                 * positions. */
-                seed = (unsigned int)rd_murmur2(&seed, sizeof(seed));
+                seed = rd_seed();
         }
-
         rand_num = rand_r(&seed);
 #else
         rand_num = rand();
 #endif
+        return rand_num;
+}
+
+#if HAVE_OSSL_SECURE_RAND_BYTES
+static int rd_rand_bytes_by_ossl(unsigned char *buf, int num) {
+        int res = -1;
+        while ((res = RAND_priv_bytes(buf, num)) == 0) {
+                rd_usleep(1000, 0); /* wait for more entropy */
+        }
+        return res;
+}
+#endif
+
+#ifdef _WIN32
+static void rd_rand_bytes_by_rand_s(unsigned char *buf, int num) {
```

Comment:
`rand_s()` is not uniformly supported in `MinGW-w64`. I think we should use `BCryptGenRandom` which is OS specific implementation and available since 2007 in all the new Windows version since then. OpenSSL uses this internally for Windows.

Reply:
Yes but we have add the additional linking to that DLL. Given both `rand_s` and `BCryptGenRandom` call `ProcessPrng` we're doing the same thing. Look [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1965199#c3) they're thinking about using `ProcessPrng` instead, but that one is available since Windows 8 while `rand_s` is since XP I'd stay with `rand_s`.